demo/c/pv_speaker_demo.c

```diff
@@ -187,38 +187,49 @@ int main(int argc, char *argv[]) {
     }
 
     fprintf(stdout, "Playing audio...\n");
-    if (pcm) {
-        char *pcmData = (char *) pcm;
-        for (int i = 0; i < num_samples; i += frame_length) {
-            bool is_last_frame = i + frame_length >= num_samples;
+    if (pcm_data) {
+        int8_t *pcm = (int8_t *) pcm_data;
+        int32_t total_written_length = 0;
 
+        while (!is_interrupted && total_written_length < num_samples) {
+            int32_t written_length = 0;
             status = pv_speaker_write(
                     speaker,
-                    is_last_frame ? num_samples - i : frame_length,
-                    &pcmData[i * bits_per_sample / 8]);
+                    &pcm[total_written_length * bits_per_sample / 8],
+                    num_samples - total_written_length,
+                    &written_length);
             if (status != PV_SPEAKER_STATUS_SUCCESS) {
                 fprintf(stderr, "Failed to write with %s.\n", pv_speaker_status_to_string(status));
                 exit(1);
             }
-            if (is_interrupted) {
-                fprintf(stdout, "\nStopped audio...\n");
-                break;
-            }
+            total_written_length += written_length;
         }
 
         free(pcm);
     }
 
+    if (is_interrupted) {
+        fprintf(stdout, "\nStopped audio...\n");
```

Comment:
I think some other mechanism for is_interrupted is needed, especially for flush. Assume you create a buffer size initially of 1 hour, you will write 1 hour to the circular buffer which will generally happen fast. when you call flush, it will be a blocking call for 1 hour. Should be the same for small second size buffers as well, currently which is 20secs, means write probably happens really fast for a 20sec audio file, flush will block 20 seconds

Reply:
added a `stop_flush` function